google/cloud/pubsublite/publisher_connection.cc

```diff
@@ -0,0 +1,215 @@
+    auto endpoint = GetEndpoint(topic.location());
+    if (!endpoint) {
+      return Status{StatusCode::kInvalidArgument, "`topic` not valid"};
+    }
+    opts.set<EndpointOption>(*std::move(endpoint));
+  }
+  std::unique_ptr<BackgroundThreads> background_threads =
+      MakeBackgroundThreadsFactory(opts)();
+  CompletionQueue cq = background_threads->cq();
+  std::shared_ptr<BackoffPolicy const> backoff_policy =
+      std::make_shared<ExponentialBackoffPolicy>(std::chrono::milliseconds{10},
+                                                 std::chrono::seconds{10}, 2.0);
+  AsyncSleeper sleeper = [cq](std::chrono::milliseconds dur) mutable {
+    return cq.MakeRelativeTimer(dur).then(
+        [](future<StatusOr<std::chrono::system_clock::time_point>> f) {
+          auto status = f.get();
+          if (!status.ok()) {
+            GCP_LOG(INFO) << "`MakeRelativeTimer` returned a non-ok status: "
+                          << status.status();
+          };
+        });
+  };
+
+  auto publisher_service_stub = CreateDefaultPublisherServiceStub(cq, opts);
+  auto batching_options = CreateBatchingOptions(opts);
+
+  std::function<std::shared_ptr<
+      Publisher<google::cloud::pubsublite::v1::Cursor>>(std::uint32_t)>
+      partition_publisher_factory = [topic, backoff_policy, batching_options,
+                                     cq, publisher_service_stub,
+                                     sleeper](std::uint32_t partition) {
+        InitialPublishRequest ipr;
+        *ipr.mutable_topic() = topic.FullName();
+        ipr.set_partition(partition);
+        AlarmRegistryImpl alarm_registry{cq};
+        auto stream_factory = MakeStreamFactory(
+            publisher_service_stub, cq, MakeClientMetadata(topic, partition));
+        return std::make_shared<PartitionPublisher>(
+            [backoff_policy, publisher_service_stub, cq, sleeper,
+             stream_factory,
+             topic](StreamInitializer<PublishRequest, PublishResponse>
+                        initializer) {
+              return absl::make_unique<ResumableAsyncStreamingReadWriteRpcImpl<
+                  PublishRequest, PublishResponse>>(
+                  [] { return absl::make_unique<StreamRetryPolicy>(); },
+                  backoff_policy, sleeper, stream_factory,
+                  std::move(initializer));
+            },
+            batching_options, std::move(ipr), alarm_registry);
+      };
+
+  AlarmRegistryImpl alarm_registry{cq};
+  PublishMessageTransformer transformer = &DefaultPublishMessageTransformer;
+  if (opts.has<PublishMessageTransformerOption>()) {
+    transformer = opts.get<PublishMessageTransformerOption>();
+  }
+  FailureHandler failure_handler = [](Status const& s) {
+    GCP_LOG(WARNING) << "Pub/Sub Lite connection failed: " << s;
+  };
+  if (opts.has<FailureHandlerOption>()) {
+    failure_handler = opts.get<FailureHandlerOption>();
+  }
+  return static_cast<std::unique_ptr<PublisherConnection>>(
```

Comment:
static_cast not needed, its an implicit conversion

Reply:
I originally had that but it didn't work. I think it's because the return type of the function is `StatusOr<std::unique_ptr<PublisherConnection>>` instead of `std::unique_ptr<PublisherConnection>` so the compiler isn't double implicitly casting to `std::unique_ptr<>` and then to `StatusOr<std::unique_ptr<>>`

Reply:
sgtm